### PR TITLE
[eclipse/xtext#1679] converted code to java

### PR DESCRIPTION
**Pull request by cdietrich**
[eclipse/xtext#1679] converted code to java
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/ContentAssistEntry.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+	 */
+	private Integer escapePosition;
+
+	/**
+	 * Additional text replacements to apply when this proposal is selected.
+	 * <p>
+	 * This property may not be supported by all editor frameworks.
+	 * </p>
+	 */
+	private final List<ReplaceRegion> textReplacements = new ArrayList<ReplaceRegion>();
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/ContentAssistEntry.java:107]: in the old class type of field and getters was arraylist.
ok to change?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalAcceptor.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		implements IIdeContentProposalAcceptor, Comparator<Pair<Integer, ContentAssistEntry>> {
+
+	protected final Set<Pair<Integer, ContentAssistEntry>> entries = new TreeSet<Pair<Integer, ContentAssistEntry>>(
+			this);
+
+	@Override
+	public void accept(ContentAssistEntry entry, int priority) {
+		if (entry != null) {
+			if (entry.getProposal() == null)
+				throw new IllegalArgumentException("proposal must not be null.");
```
Comment (tivervac) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalAcceptor.java:34]: s/proposal/Proposal/

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalCreator.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import org.eclipse.xtext.xbase.lib.StringExtensions;
+
+/**
+ * Factory for content assist entries. Whenever possible, you should use this creator instead of building entries
+ * directly, since prefix matching and conflict handling is done here.
+ * 
+ * @since 2.10
+ * @noreference
+ */
+@SuppressWarnings("all")
```
Comment (tivervac) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalCreator.java:26]: Probably want to remove this

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalCreator.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+			if (init != null) {
+				init.apply(result);
+			}
+			return result;
+		}
+		return null;
+	}
+
+	public boolean isValidProposal(String proposal, String prefix, ContentAssistContext context) {
+		return !StringExtensions.isNullOrEmpty(proposal) && prefixMatcher.isCandidateMatchingPrefix(proposal, prefix)
```
Comment (tivervac) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalCreator.java:92]: You can just use `Strings.isNullOrEmpty` here

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeCrossrefProposalProvider.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import com.google.inject.Inject;
+
+/**
+ * Special content assist proposal provider for cross-references.
+ * 
+ * @noreference
+ */
+public class IdeCrossrefProposalProvider {
+
+	private static Logger LOG = Logger.getLogger(IdeCrossrefProposalProvider.class);
```
Comment (tivervac) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeCrossrefProposalProvider.java:27]: final?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+/**
+ * Maps an EObject to the corresponding {@link DocumentSymbol document symbol}.
+ */
+@Beta
+@Singleton
+public class DocumentSymbolMapper {
+	/**
+	 * Provides {@link DocumentSymbol#detail detail} for a {@link DocumentSymbol document symbol}.
+	 * <p>
+	 * Always returns with an empty string by default.
```
Comment (tivervac) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java:37]: Always returns an empty string by default*

**Comment by cdietrich** [comment]
see also https://github.com/eclipse/xtext-extras/pull/588

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/ContentAssistEntry.java (hunk 1 above)
Comment (tivervac) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/ContentAssistEntry.java:107]: If it's only the field, not public getter, it should be fine

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	@Test
+	public void withoutNull() {
+		final List<? extends DocumentHighlight> input = sort(
+				Lists.newArrayList(newHighlight(DocumentHighlightKind.Text, newRange(2, 2, 2, 2)),
+						newHighlight(DocumentHighlightKind.Text, newRange(1, 1, 1, 1)),
+						newHighlight(DocumentHighlightKind.Write, newRange(2, 2, 2, 2)),
+						newHighlight(DocumentHighlightKind.Write, newRange(1, 1, 1, 1)),
+						newHighlight(DocumentHighlightKind.Read, newRange(2, 2, 2, 2)),
+						newHighlight(DocumentHighlightKind.Read, newRange(1, 1, 1, 1))));
+		Assert.assertEquals(1, input.get(0).getRange().getStart().getLine());
```
Comment (nbhusare) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java:53]: The class extends from `Assert`, you can just use `assertEquals`.

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		Assert.assertEquals(DocumentHighlightKind.Write, input.get(5).getKind());
+	}
+
+	@Test
+	public void withNull() {
+		final List<? extends DocumentHighlight> input = sort(
+				Lists.newArrayList(null, newHighlight(DocumentHighlightKind.Text, newRange(1, 1, 1, 1)),
+						newHighlight(DocumentHighlightKind.Write, newRange(1, 1, 1, 1)),
+						newHighlight(DocumentHighlightKind.Read, newRange(1, 1, 1, 1))));
+		Assert.assertEquals(1, input.get(0).getRange().getStart().getLine());
```
Comment (nbhusare) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java:91]: PS comment above.

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		Assert.assertEquals(DocumentHighlightKind.Read, input.get(1).getKind());
+		Assert.assertEquals(1, input.get(2).getRange().getStart().getLine());
+		Assert.assertEquals(1, input.get(2).getRange().getStart().getCharacter());
+		Assert.assertEquals(1, input.get(2).getRange().getEnd().getLine());
+		Assert.assertEquals(1, input.get(2).getRange().getEnd().getCharacter());
+		Assert.assertEquals(DocumentHighlightKind.Write, input.get(2).getKind());
+		Assert.assertNull(IterableExtensions.last(input));
+	}
+
+	private DocumentHighlight newHighlight(final DocumentHighlightKind kind, final Range range) {
```
Comment (nbhusare) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java:109]: Please remove `final`. Kindly do the same on lines 113, 118 too.

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/RangeComparatorTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+@InjectWith(TestLanguageIdeInjectorProvider.class)
+public class RangeComparatorTest extends Assert {
+	@Inject
+	private RangeComparator comparator;
+
+	@Test
+	public void withoutNull() {
+		List<? extends Range> input = sort(Lists.newArrayList(newRange(1, 2, 1, 2), newRange(1, 1, 2, 1),
+				newRange(1, 1, 1, 2), newRange(1, 1, 1, 1), newRange(2, 2, 2, 3)));
+		Assert.assertEquals(1, input.get(0).getStart().getLine());
```
Comment (nbhusare) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/RangeComparatorTest.java:40]: `Assert.assertEquals(...)` -> `assertEquals(...)`

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/RangeComparatorTest.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		Assert.assertEquals(2, input.get(4).getStart().getLine());
+		Assert.assertEquals(2, input.get(4).getStart().getCharacter());
+		Assert.assertEquals(2, input.get(4).getEnd().getLine());
+		Assert.assertEquals(3, input.get(4).getEnd().getCharacter());
+	}
+
+	@Test
+	public void withNull() {
+		List<? extends Range> input = sort(Lists.newArrayList(newRange(2, 2, 2, 3), null, newRange(1, 1, 1, 1)));
+		Assert.assertEquals(1, input.get(0).getStart().getLine());
```
Comment (nbhusare) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/RangeComparatorTest.java:65]: PS above comment.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalAcceptor.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+			return s1.compareTo(s2);
+		}
+		return ignoreCase;
+	}
+
+	public Iterable<ContentAssistEntry> getEntries() {
+		return Iterables.transform(entries, e -> e.getValue());
+	}
+
+}
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeContentProposalAcceptor.java:62]: NL

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeCrossrefProposalProvider.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+			ContentAssistContext context) {
+		return proposalCreator.createProposal(qualifiedNameConverter.toString(candidate.getName()), context, (e) -> {
+			e.setSource(candidate);
+			e.setDescription(candidate.getEClass() != null ? candidate.getEClass().getName() : null);
+			e.setKind(ContentAssistEntry.KIND_REFERENCE);
+		});
+	}
+
+	protected IQualifiedNameConverter getQualifiedNameConverter() {
+		return this.qualifiedNameConverter;
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeCrossrefProposalProvider.java:70]: `this.qualifiedNameConverter` -> `qualifiedNameConverter`

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeCrossrefProposalProvider.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	protected IdeContentProposalCreator getProposalCreator() {
+		return this.proposalCreator;
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/IdeCrossrefProposalProvider.java:74]: `return this.proposalCreator` -> `return proposalCreator`

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+	public boolean isRecursive() {
+		if (recursive == null) {
+			recursive = Wrapper.wrap(internalIsRecursive());
+		}
+		return recursive.get().booleanValue();
+	}
+
+	protected boolean internalIsRecursive() {
+		IHierarchyNode node = parent;
+		while (node != null) {
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.java:54]: The indentation of the code inside the while-loop looks incorrect. Please confirm.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+					return true;
+				}
+				node = node.getParent();
+		}
+		return false;
+	}
+
+	@Override
+	public boolean mayHaveChildren() {
+		return this.mayHaveChildren;
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyNode.java:67]: `this.mayHaveChildren` -> `mayHaveChildren`

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyRoot.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * @author kosyakov - Initial contribution and API
+ * @since 2.10
+ */
+public class DefaultHierarchyRoot implements IHierarchyRoot {
+	private final List<IHierarchyNode> roots = new ArrayList<>();
+
+	public List<IHierarchyNode> getRoots() {
+		return this.roots;
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/hierarchy/DefaultHierarchyRoot.java:22]: `this.roots` -> `roots`

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+		if (objectRange != null) {
+			documentSymbol.setRange(objectRange);
+		}
+		Range objectSelectionRange = rangeProvider.getSelectionRange(object);
+		if (objectSelectionRange != null) {
+			documentSymbol.setSelectionRange(objectSelectionRange);
+		}
+		documentSymbol.setDetail(detailsProvider.getDetails(object));
+		documentSymbol.setDeprecated(deprecationInfoProvider.isDeprecated(object));
+		documentSymbol.setChildren(new ArrayList<>());
```
Comment (nbhusare) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java:189]: `new ArrayList<>()` -> `Collections.emptyList()`

**Comment by nbhusare** [comment]
Minor nits and LGTM

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/ContentAssistEntry.java (hunk 1 above)
Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/contentassist/ContentAssistEntry.java:107]: yes but there is a noreference on the class

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java (hunk 18 above)
Comment (tivervac) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java:189]: I think elements are added to it later on which will cause exceptions if it's the empty list

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolMapper.java:189]: i did not want to change the semantics here thus kept the list

**Comment by cdietrich** [comment]
@tivervac please reexamine

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.runner.RunWith;
+
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+
+/**
+ * Test for the null-safe {@link DocumentHighlightComparator}.
+ * 
+ * <p>
+ * This test focuses on the {@link DocumentHighlightKind} and the null-safety
```
Comment (tivervac) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java:33]: safety.

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+ * other orderings are checked via the {@link RangeComparatorTest}.
```
Comment (tivervac) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/util/DocumentHighlightComparatorTest.java:34]: Other

**Pull request merged**